### PR TITLE
REF: Move most remaining arith helpers to datetimelike mixins

### PR DESCRIPTION
**Pull request by jbrockmendel**
Before we can finally move over `_add_datetimelike_methods` and be done with this, some more work needs to go into the `__new__` methods.  This gets another big chunk through, starts implementing tests.

### CODE REVIEW

File: pandas/core/arrays/period.py [hunk 1]
```diff
@@ -113,12 +115,23 @@ def freq(self, value):
 
     _attributes = ["freq"]
 
+    def __new__(cls, values, freq=None, **kwargs):
```
Comment (jorisvandenbossche) [inline pandas/core/arrays/period.py]: maybe add a todo comment to convert this to init once the Mixin is converted to proper array

**Review by jorisvandenbossche — APPROVED** [review]

File: pandas/core/arrays/datetimelike.py [hunk 2]
```diff
@@ -108,38 +113,43 @@ def __getitem__(self, key):
...
-                # To support MPL which performs slicing with 2 dim
-                # even though it only has 1 dim by definition
-                if is_period:
-                    return self._simple_new(result, **attribs)
-                return result
+        return self._simple_new(result, **attribs)
 
-            return self._simple_new(result, **attribs)
+    def astype(self, dtype, copy=True):
+        if is_object_dtype(dtype):
```
Comment (jreback) [inline pandas/core/arrays/datetimelike.py]: at some point pls add a doc-string

File: pandas/tests/arrays/test_datetimelike.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        tz = tz_naive_fixture
+        dti = pd.date_range('2016-01-01', periods=3, tz=tz)
+        arr = DatetimeArrayMixin(dti)
+        asobj = arr.astype('O')
+        assert isinstance(asobj, np.ndarray)
+        assert asobj.dtype == 'O'
+        assert list(asobj) == list(dti)
+
+
+class TestTimedeltaArray(object):
```
Comment (jreback) [inline pandas/tests/arrays/test_datetimelike.py]: k for now, maybe split to sub-dirs at some point

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
looks ok to me. merge on green.

**Comment by jreback** [comment]
thanks @jbrockmendel

**Pull request merged**